## Scripts
This folder contains all custom scripts created to complete this project, including a general usage description. 

# blast.sh
This script all sequences in the folder "fasta_resultaten" against a pre-specified target genome and reports all findings to the directory "blast_resultaten". 
input: 
1. A directory "fasta_resultaten" in the directory where blast.sh runs, with at least 1 fasta file with a single sequence. 
2. A fasta file for the target genome in the same directory where blast.sh runs. 
output:
1. Tabular (version 7) BLAST results, one file for each sequence in "fasta_resultaten", stored in the directory "blast_resultaten".

# Slice script
This script can run in either of two modes. Mode 1 slices all positions from a given GFF3-file and it's related fasta genome file. 
- input mode 1:
1. genome of species 1 in fasta format.
2. gff3 file related to genome given at input 1.
3. Newline saperated list containing only gene IDs of interest. 
- output mode 1:
One fasta file for every feature in the gff3 record at 1, containing the sequence for that feature for the reference species. The filenames follow the following naming convention:
- geneId:feature_type.fa
Running this script in it's second modus requires a file containing BLAST results. These BLAST results and knowledge of the query sequence is used which result is most likely its counterpart on the secondary parents genome. A fasta file with the secondary parents genome is also required.
- input mode 2:
1. One text file containing BLAST results in the following format:
filename_of_query_fasta \t contig_of_query_sequence \t query_start \t query_stop \n
2. genome of species 2 in fasta format. 
- output mode 2:
Updates on all fasta files in directory "fasta_resultaten", appending every sequence of the second parent to the fasta file of the first parent. 

# Shortlist_script.sh
**PENDING INPUT BY NINO**

# alignments.sh
**TO BE CONFIRMED BY NINO**
This script aligns all feature files generated by the slice script, using the programm MUSCLE.
- Input:
1. Any number of fasta files in a separate directory.
2. One aligned output file for every input file, outputted to a different (?) directory. 

# gff3_update.py
**PENDING INPUT BY NINO**

# snp_caller.py
A script used to determine SNPs by type and count from the combination of nucleotide, aminoacid, fasta, and MUSCLE files. As of right now this documentation is incomple, as snp_caller.py is still subject to major changes. 


