## Scripts
This folder contains all custom scripts created to complete this project, including a general usage description. 

# Slice script
This script slices the DNA-sequence of features in a GFF3-file from the corresponding genome fasta file, creating a saperate fasta file for every feature, containing the sequence for the reference genome and the Jersey kale.
- input:
1. genome of species 1 in fasta format.
2. gff3 file related to genome given at input 1.
3. genome of species 2 in fasta format.
4. gff3 file related to genome given at input 3.
- output:
Two fasta files for every feature in the gff3 record at 1, containing the sequence for that feature for both species. The filenames follow the following naming convention:
1. nucleotide: featureType:geneId:featureStart:featureStop.fa
2. amino acid: featureType:geneId:featureStart:featureStop.aa

# Shortlist_script.sh
**PENDING**

# alignment.sh
This script aligns all feature files generated by the slice script, using the programm MUSCLE.
- Input:
1. Any number of fasta files in a separate directory with the following naming convention: Feature\[count\].afa
- Output:
2. One aligned output file for every input file, with the same name and a new extention '.afa'.

# gff3_update.py
**PENDING**

# shortlist_sort.py
A script used to prepare a GFF3-file for further use over the course of this project. It takes a shortlist variant of a GFF3-file (only containing genes of interest) as input. For every gene, all gene features are gathered, sorted in transcribing order, given NonCoding Region records, and sorted again in transcribing order. This ensures  the slicing script (and counting scripts later on) have an easy way of counting silent SNPs (all SNPs occuring on NCR records) and determining transcribing order (which is now simply top to bottom in the GFF3 file).

Input: as the script has not been given an argument parser yet, it has fixed names for input- and output files.
1. testgff3_01.txt. Any GFF3 formatted file with this specific name.
Output:
1. shortlist.gffex. A file following GFF3-formatting, except it also contains records for every noncoding region, and features are grouped together by gene ID instead of sorting the list by order of start position from left to right. 


# snp_caller.py
A script used to 


